### DISCUSSION TITLE
Should facts be distinct by type, are "general" facts encouraged or discouraged?

### DISCUSSION THREAD

**Discussion by douglasg14b** [NRules/NRules#335] [Q&A]
Is there an opinion, either for performance, semantics, patterns of use or otherwise. That facts be distinct by type, or should they be "general"?

I'm probably describing this horribly because words are hard, so before you interpret that question, let me provide an example:

```cs
// Each "fact" has a type that semantically represents the entirety of a self-contained concept
// Rules would be more sparse, matching on very few facts
public record MessageFact(string MessageId, DateTimeOffset Timestamp) {}
public record AgentMessageFact(string MessageId, DateTimeOffset Timestamp) {}
public record CustomerMessageFact(string MessageId, DateTimeOffset Timestamp) {}
public record AgentMessageAfterAssignmentFact(string MessageId, DateTimeOffset Timestamp) {}
public record CustomerMessageAfterAssignmentFact(string MessageId, DateTimeOffset Timestamp) {}
// ...etc with various other mutations of a message, some temporal, some categorial

// Facts can instead be more "general purpose", with each fact representing a combination/set of possible facts
// Rules would be more general, potentially more complex, matching on more values

public enum MessageSender
{
    Customer,
    Expert,
    Bot,
    Unknown
}

public record MessageFact(string MessageId, DateTimeOffset Timestamp, MessageSender Sender) { }
public record MessageAfterAssignmentFact(string MessageId, DateTimeOffset Timestamp) {}
```

**Are there upsides & downsides to doing this one way or the other? Do you have an opinion on this sort of utilization of your library?**

<hr>

Instead of just asking this openly, I tried to consider some for myself, perhaps you can (in)validate these in addition to sharing your opinion.

**Each "fact" has a type that is semantically after that fact:**
- 🟢 Rules are relatively straight forward to operate, and more comprehendable
- 🟢 Facts clearly indicate what they are, and nothing more
- 🟢 Fact names will more often 1-to-1 represent domain concepts
- 🔴 Mapping distinct facts back into domain constants (such as a `MessageSender` enum) when such an constant is required
- 🔴 There are many rules & facts, it may become easy to get lost as the quantity expands
- 🔴 There are many more facts, perhaps this represents a performance concern

**More general purpose facts:**
- 🟢 Fewer facts & rules may be easier to maintain & visually grep through
- 🟢 Facts can make use of domain constants and concepts directly without need for a translation/adapter layer
- 🔴 Facts require additional logic to interpret them
- 🔴 A fact by it's existence may not represent a self contained domain concept (`Message where Sender = "Expert"` vs `CustomerMessage`)

**Comment by snikolayev** [comment]
@douglasg14b from the performance perspective, I think this decision is insignificant, because fact type is simply yet another condition the engine is using to filter down facts in the Rete graph. All facts first go to a TypeNode that routes facts by their type to the corresponding subtree in the graph. So, matching an ExpertMessage will first go to a TypeNode where Type = ExpertMessage, whereas a Message { Sender = "Expert" } will go to a TypeNode where Type = Message and then to a SelectNode where Sender = "Expert". I think any performance differences here will be dwarfed by whatever else you do in your rules, e.g. how selective your match conditions are, the magnitude of joins, etc.
The second part of the question, regarding semantics, is harder to answer. I think this is not even that specific to rules. A very similar debate exists between a DDD approach of a rich domain model and an anemic domain model with various service methods.
I think the domain model and rules should be at the level that best facilitates iterative implementation of your rules. Meaning, it should lead to self-contained composable rules as much as possible. Also, consider polymorphic rules and facts - e.g. ExpertMessage fact inheriting from a Message fact, where you can write specific rules matching the ExpertMessage, but can also write generic rules matching any Message.

**Comment by snikolayev** [comment]
FWIW I personally prefer more specific facts, closer tied to real concepts in my domain in a 1:1 way
